File: packages/agent-harness-tools/src/log-stream.ts
import nodeFs from "node:fs";
import type { FSWatcher } from "node:fs";
import { hasOwnErrorCode } from "./error-codes.js";
import type { LogChunk } from "./execution-env.js";

const JOB_DIR = "/tmp/poe-jobs";
const POLL_INTERVAL_MS = 250;

export interface LogStreamEnv {
  fs?: LogStreamFs;
}

export interface LogStreamFs {
  promises: {
    readFile(path: string): Promise<Buffer | string>;
    stat?(path: string): Promise<{ mtimeMs: number }>;
    lstat?(path: string): Promise<{ isSymbolicLink(): boolean }>;
  };
  watch?: (path: string, listener: () => void) => FSWatcher;
}

export function wrapForLogTee(argv: string[], jobId: string): string[] {
  assertSafeJobId(jobId);

  if (argv.length === 0) {
    throw new Error("wrapForLogTee requires argv to contain at least one argument");
  }

  const command = argv.map(shellQuote).join(" ");
  const logFile = shellQuote(jobLogPath(jobId));
  const exitFile = shellQuote(jobExitPath(jobId));
  const exitTmpFile = shellQuote(`${jobExitPath(jobId)}.tmp`);
  const safetyChecks = [
    `test ! -L ${shellQuote(JOB_DIR)}`,
    `test ! -L ${logFile}`,
    `test ! -L ${exitFile}`,
    `test ! -L ${exitTmpFile}`
  ].join(" && ");
  const script = [
    `mkdir -p ${shellQuote(JOB_DIR)}`,
    safetyChecks,
    `({ (${command}); echo $? > ${exitTmpFile}; } 2>&1 | tee ${logFile}; mv ${exitTmpFile} ${exitFile})`
  ].join(" && ");

  return ["sh", "-c", script];
}

export async function* streamLogFile(
  env: LogStreamEnv,
  jobId: string,
  opts: { sinceByte?: number; since?: Date; follow?: boolean }
): AsyncIterable<LogChunk> {
  assertSafeJobId(jobId);

  const fs = env.fs ?? nodeFs;
  const file = jobLogPath(jobId);
  let byteOffset = opts.sinceByte ?? (opts.since === undefined ? 0 : await readCurrentByteLength(fs, file));
  let pendingBytes: Buffer = Buffer.alloc(0);
  let pendingByteOffset = byteOffset;

  while (true) {
    if (opts.since !== undefined && !(await wasModifiedSince(fs, file, opts.since))) {
      await waitForLogChange(fs, file);
      continue;
    }

    const result = await readLogChunk(fs, file, byteOffset);
    if (result !== null) {
      const combined = pendingBytes.length === 0
        ? result.bytes
        : Buffer.concat([pendingBytes, result.bytes]);
      const completeLength = completeUtf8PrefixLength(combined);
      byteOffset = result.nextByteOffset;
      pendingBytes = combined.subarray(completeLength);
      const data = combined.subarray(0, completeLength).toString("utf8");
      if (data.length > 0) {
        yield { byteOffset: pendingByteOffset, data };
        pendingByteOffset += completeLength;
      }
      continue;
    }

    if ((await readTextFileIfExists(fs, jobExitPath(jobId))) !== null) {
      return;
    }

    if (opts.follow === false) {
      return;
    }

    await waitForLogChange(fs, file);
  }
}

async function readCurrentByteLength(fs: LogStreamFs, file: string): Promise<number> {
  const contents = await readFileIfExists(fs, file);
  return contents?.byteLength ?? 0;
}

async function wasModifiedSince(fs: LogStreamFs, file: string, since: Date): Promise<boolean> {
  if (fs.promises.stat === undefined) {
    return true;
  }

  try {
    const stat = await fs.promises.stat(file);
    return stat.mtimeMs >= since.getTime();
  } catch (error) {
    if (hasOwnErrorCode(error, "ENOENT")) {
      return false;
    }
    throw error;
  }
}

export async function waitForExit(
  env: LogStreamEnv,
  jobId: string,
  opts: { signal?: AbortSignal } = {}
): Promise<{ exitCode: number }> {
  assertSafeJobId(jobId);

  const fs = env.fs ?? nodeFs;
  const file = jobExitPath(jobId);

  while (true) {
    throwIfAborted(opts.signal);
    const contents = await readTextFileIfExists(fs, file);
    if (contents !== null) {
      const text = contents.trim();
      const exitCode = Number(text);
      if (text.length === 0 || !Number.isInteger(exitCode)) {
        throw new Error(`Invalid exit code in ${file}: ${contents}`);
      }
      return { exitCode };
    }

    await sleep(POLL_INTERVAL_MS, opts.signal);
  }
}

function jobLogPath(jobId: string): string {
  return `${JOB_DIR}/${jobId}.log`;
}

function jobExitPath(jobId: string): string {
  return `${JOB_DIR}/${jobId}.exit`;
}

function assertSafeJobId(jobId: string): void {
  if (jobId.length === 0 || jobId === "." || jobId === ".." || jobId.includes("/") || jobId.includes("\\")) {
    throw new Error(`Invalid job id "${jobId}". Job ids must be single filename components.`);
  }
}

async function readLogChunk(
  fs: LogStreamFs,
  file: string,
  byteOffset: number
): Promise<{ bytes: Buffer; nextByteOffset: number } | null> {
  const contents = await readFileIfExists(fs, file);
  if (contents === null || byteOffset >= contents.byteLength) {
    return null;
  }

  return {
    bytes: contents.subarray(byteOffset),
    nextByteOffset: contents.byteLength
  };
}

async function readTextFileIfExists(fs: LogStreamFs, file: string): Promise<string | null> {
  const contents = await readFileIfExists(fs, file);
  return contents?.toString("utf8") ?? null;
}

async function readFileIfExists(fs: LogStreamFs, file: string): Promise<Buffer | null> {
  try {
    await assertRegularManagedFile(fs, file);
    const contents = await fs.promises.readFile(file);
    return Buffer.isBuffer(contents) ? contents : Buffer.from(contents);
  } catch (error) {
    if (hasOwnErrorCode(error, "ENOENT")) {
      return null;
    }
    throw error;
  }
}

async function assertRegularManagedFile(fs: LogStreamFs, file: string): Promise<void> {
  const lstat = fs.promises.lstat;
  if (lstat === undefined) {
    return;
  }

  await assertManagedJobDirectory(lstat);

  try {
    if ((await lstat(file)).isSymbolicLink()) {
      throw new Error("Managed job file must not be a symbolic link.");
    }
  } catch (error) {
    if (hasOwnErrorCode(error, "ENOENT")) {
      return;
    }

    throw error;
  }
}

async function assertManagedJobDirectory(
  lstat: NonNullable<LogStreamFs["promises"]["lstat"]>
): Promise<void> {
  try {
    if ((await lstat(JOB_DIR)).isSymbolicLink()) {
      throw new Error("Managed job directory must not be a symbolic link.");
    }
  } catch (error) {
    if (hasOwnErrorCode(error, "ENOENT")) {
      return;
    }

    throw error;
  }
}

function completeUtf8PrefixLength(contents: Buffer): number {
  if (contents.length === 0) {
    return 0;
  }

  let leadIndex = contents.length - 1;
  while (leadIndex >= 0 && isUtf8ContinuationByte(contents[leadIndex]!)) {
    leadIndex -= 1;
  }

  if (leadIndex < 0) {
    return contents.length;
  }

  const expectedLength = utf8SequenceLength(contents[leadIndex]!);
  if (expectedLength === 0) {
    return contents.length;
  }

  const availableLength = contents.length - leadIndex;
  return availableLength < expectedLength ? leadIndex : contents.length;
}

function isUtf8ContinuationByte(byte: number): boolean {
  return byte >= 0x80 && byte <= 0xbf;
}

function utf8SequenceLength(byte: number): number {
  if (byte >= 0xc2 && byte <= 0xdf) {
    return 2;
  }
  if (byte >= 0xe0 && byte <= 0xef) {
    return 3;
  }
  if (byte >= 0xf0 && byte <= 0xf4) {
    return 4;
  }
  return 0;
}

async function waitForLogChange(fs: LogStreamFs, file: string): Promise<void> {
  const watch = fs.watch;
  if (typeof watch !== "function") {
    await sleep(POLL_INTERVAL_MS);
    return;
  }

  await new Promise<void>((resolve) => {
    let watcher: FSWatcher | null = null;
    const timer = setTimeout(done, POLL_INTERVAL_MS);

    function done(): void {
      clearTimeout(timer);
      watcher?.close();
      resolve();
    }

    try {
      watcher = watch(file, done);
    } catch {
      done();
    }
  });
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    let timer: NodeJS.Timeout | null = null;
    const abort = () => {
      if (timer !== null) {
        clearTimeout(timer);
      }
      reject(new Error("waitForExit aborted."));
    };

    if (signal?.aborted) {
      abort();
      return;
    }

    timer = setTimeout(() => {
      signal?.removeEventListener("abort", abort);
      resolve();
    }, ms);
    signal?.addEventListener("abort", abort, { once: true });
  });
}

function throwIfAborted(signal: AbortSignal | undefined): void {
  if (signal?.aborted) {
    throw new Error("waitForExit aborted.");
  }
}

function shellQuote(value: string): string {
  return `'${value.replaceAll("'", "'\\''")}'`;
}

File: packages/agent-harness-tools-rust/src/log_stream.rs
//! Managed log admission, byte framing and portable shell quoting.
pub fn safe_job_id(id: &[u16]) -> bool {
    !id.is_empty()
        && id != [46]
        && id != [46, 46]
        && !id.iter().any(|unit| matches!(*unit, 0 | 47 | 92))
}
pub fn complete_utf8_prefix(bytes: &[u8]) -> usize {
    let Some(lead) = bytes.iter().rposition(|byte| !matches!(*byte, 0x80..=0xbf)) else {
        return bytes.len();
    };
    let expected = match bytes[lead] {
        0xc2..=0xdf => 2,
        0xe0..=0xef => 3,
        0xf0..=0xf4 => 4,
        _ => 0,
    };
    if expected > 0 && bytes.len() - lead < expected {
        lead
    } else {
        bytes.len()
    }
}
pub fn decimal_exit_code(value: &[u16]) -> Option<f64> {
    if value == [48] {
        return Some(0.0);
    }
    if !matches!(value.first(), Some(49..=57)) || !value.iter().all(|unit| matches!(*unit, 48..=57))
    {
        return None;
    }
    value
        .iter()
        .map(|unit| (*unit as u8) as char)
        .collect::<String>()
        .parse()
        .ok()
}
fn quote(value: &[u16]) -> Vec<u16> {
    let mut out = vec![39];
    for unit in value {
        if *unit == 39 {
            out.extend([39, 92, 39, 39]);
        } else {
            out.push(*unit);
        }
    }
    out.push(39);
    out
}
pub fn tee_command(argv: &[Vec<u16>], job: &[u16]) -> Result<Vec<Vec<u16>>, &'static str> {
    if !safe_job_id(job) {
        return Err("Invalid job id");
    }
    if argv.is_empty() {
        return Err("wrapForLogTee requires argv to contain at least one argument");
    }
    let mut log: Vec<u16> = "/tmp/poe-jobs/".encode_utf16().collect();
    log.extend(job);
    let mut exit = log.clone();
    exit.extend(".exit".encode_utf16());
    log.extend(".log".encode_utf16());
    let mut tmp = exit.clone();
    tmp.extend(".tmp".encode_utf16());
    let log = quote(&log);
    let exit = quote(&exit);
    let tmp = quote(&tmp);
    let mut script: Vec<u16> = "mkdir -p '/tmp/poe-jobs' && test ! -L '/tmp/poe-jobs'"
        .encode_utf16()
        .collect();
    for target in [&log, &exit, &tmp] {
        script.extend(" && test ! -L ".encode_utf16());
        script.extend(target);
    }
    script.extend(" && ({ (".encode_utf16());
    for (index, arg) in argv.iter().enumerate() {
        if index > 0 {
            script.push(32);
        }
        script.extend(quote(arg));
    }
    script.extend("); echo $? > ".encode_utf16());
    script.extend(&tmp);
    script.extend("; } 2>&1 | tee ".encode_utf16());
    script.extend(log);
    script.extend("; mv ".encode_utf16());
    script.extend(tmp);
    script.push(32);
    script.extend(exit);
    script.extend(")".encode_utf16());
    Ok(vec![
        "sh".encode_utf16().collect(),
        "-c".encode_utf16().collect(),
        script,
    ])
}
